src/Build/BackEnd/Components/ProjectCache/CacheResult.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Experimental.ProjectCache
+{
+    public enum CacheResultType
+    {
+        CacheHit,
+        CacheMiss,
+        CacheNotApplicable,
+        CacheError
```

Comment:
Is it worth keeping `CacheError`? I'm thinking of dropping it in favor of just checking `PluginLoggerBase.HasLoggedErrors`. If we keep both, what should the user or msbuild do if the plugin returns `CacheError` but no errors are logged? Seems simpler to just keep one mechanism for signaling errors.

Reply:
Yeah I agree, the way its mostly done is its success unless an error was logged.

Reply:
K, I'll remove CacheError

Reply:
So it turns out CacheError may have a good use: when the plugin is queried for a project but it encounters and logs one or more errors, what should the plugin return? It could either return a null CacheResult or a CacheResult of result type CacheError (returning a CacheMiss or CacheNotApplicable doesn't seem a good fit for the error case). But then there's the above ambiguity when a plugin could return CacheError without logging any errors.

An alternative is to remove the enum entirely. On cache hits the plugin returns a CacheResult with the build results (or proxy targets and whatnot). On anything else (cache miss, cache not applicable, cache error) the plugin returns a null CacheResult and logs why it couldn't satisfy the request. Worst case we get a plugin that returns null and does not log anything.

I am inclined now to keep CacheError in order to make the return modes explicit.

Let me know what you think.

Reply:
Decided to just replace `CacheError` with a `None` value. Hit, Miss, and NA are domain level concepts that sort of define the meaning of a plugin so I like to have them explicit. Without them, we'd rely on arbitrary plugin implementations on how they express these 3 possibilities, versus MSBuild giving a standardized message for each type of response. Another concrete use case for having them is to enable MSBuild to report the cache hit ratio in the build summary (`hits / (hits + misses)`, need NA to not count them as misses). That's a very useful metric for cacheability health (which, just like perf, is something that tends to degrade over time).